### PR TITLE
race, storage:  add independent `AddNames` and `RemoveNames` for images,layers and containers

### PR DESCRIPTION
**Pull request by flouthoc**
Adds `SetNameWithOptions` so operations which are invoke in parallel
manner can use it without destroying names from storage.

For instance

We are deleting names which were already written in store.
This creates faulty behavior when builds are invoked in parallel manner, as
this removes names for other builds.

To fix this behavior we must append to already written names and
override if needed. But this should be optional and not break public API

Following patch will be used by parallel operations at podman or buildah end, directly or indirectly.

PR replaces: https://github.com/containers/storage/pull/1152

### CODE REVIEW

File: containers.go
Comment (vrothberg) [inline containers.go:87]: Still a breaking change.

**Review by vrothberg — COMMENTED** [review]
Would a `{Add,Delete}Name` API address the races?

File: containers.go
Comment (giuseppe) [inline containers.go:391]: I think you can drop the first dedupe at line 381 and just do it here not guarded by `opts.AppendOnly`.

File: containers.go
Comment (flouthoc) [reply containers.go]: @vrothberg Yeah i was not aware that container store is exposed dropped it.

File: containers.go
Comment (flouthoc) [reply containers.go]: yes. that would be better.

**Comment by flouthoc** [comment]
> Would a `{Add,Delete}Name` API address the races?

@vrothberg Bunch of code in `common or podman or buildah` needs to be instrumented slowly. So this seems most intuitive from what we discussed here   https://github.com/containers/storage/pull/1152 . 

There are some other things as well where we might have race but maybe that is part of a different PR.

**Review by giuseppe — APPROVED** [review]
code changes LGTM

**Comment by giuseppe** [comment]
> Would a {Add,Delete}Name API address the races?

Wouldn't that be more difficult to handle for the caller?  If it is adding multiple names and the operation fails middle way, should it then try to delete previously added ones?

File: containers.go
Comment (mtrmac) [inline containers.go:402]: This looks quite broken: Not deleting `container` from `r.byname` will set `otherContainer` to `container` AFAICS.

I think that’s ultimately hidden by the `container.Names` assignment below, but it’s just not a clean and consistent way to implement this.

---

Same in the other two implementations.

---

One way to do this would be not to change the basic structure, only to change what names are set. I.e. keep the original `delete(r.byname)` loop as is, _then_ compute the final `names` value, and then proceed with the `otherContainer/byname/…` updates using the existing logic.

That would have the advantage that `updatedNames := applyEdits(container.Names, inputNames, opts)` could be shared across the three implementations.

Another way is to indeed deeply embed the add/remove cases into the implementation, so that names that don’t change are neither removed nor added in `r.byname`. That’s _technically_ more efficient but IMHO not worth the complexity.

**Review by mtrmac — COMMENTED** [review]
If we need atomic name additions, we _certainly_ also need atomic name removals. So I’d expect something like  `UpdateNames(id string, updates NameUpdates)` with `type NameUpdates{ struct addNames, removeNames []string }`

There was also an issue I can’t quickly find about `store.CreateImage` doing name updates and other changes atomically with other operations, that’s why I suggest a `type NameUpdates` that could be reused in that other feature.

**Comment by mtrmac** [comment]
> @vrothberg Bunch of code in `common or podman or buildah` needs to be instrumented slowly.

What difference does that make? Every line calling `SetNames` should to be edited either way, whatever form we design for the new API.

**Review by mtrmac — COMMENTED** [review]
.

**Comment by mtrmac** [comment]
Also, if we know the old APIs are racy, they should be documented as deprecated.

**Comment by vrothberg** [comment]
> > @vrothberg Bunch of code in `common or podman or buildah` needs to be instrumented slowly.
> 
> What difference does that make? Every line calling `SetNames` should be edited either way, whatever form we design for the new API.

Yes, `SetNames` should not be used anymore. IIUIC, the race is in reading all data, followed by writing all data. An add(), delete() could just take care of synchronization and callers don't have to worry.

**Comment by mtrmac** [comment]
> There was also an issue I can’t quickly find about `store.CreateImage` doing name updates and other changes atomically with other operations

https://github.com/containers/storage/issues/595

File: containers.go
Comment (mtrmac) [reply containers.go:402]: … actually, with a clean `addNames` / `removeNames` (were we do need to handle no-op names but they are only no-op additions / no-op removals) the “deeply embed” approach also might be simple enough.

File: store.go
Comment (nalind) [inline store.go:2065]: This is not a valid assumption, and could break untagging unless callers switch to `SetNamesWithOptions`.  An `AddNames/RemoveNames` pair instead of `SetNamesWithOptions()` would avoid this.

File: containers.go
Comment (flouthoc) [reply containers.go:401]: Yeah this part is not needed. Let me alter that, lets discuss new in original thread.

**Comment by flouthoc** [comment]
If we have a consensus on new API then i guess its cool. We can create a new API.

File: containers.go
Comment (vrothberg) [inline containers.go:407]: I'd much prefer to error out early:
```go
container, found := r.lookup(id)
if !found {
    return ErrContainer...
}
// logic
```

File: containers.go
Comment (vrothberg) [inline containers.go:90]: I am not a native speaker but something similar to the comment below easier to brain parse:
```suggestion
	// AddNames adds the supplied values to the list of names associated with the container with the specified id.
```
Same at other locations.

File: containers.go
Comment (vrothberg) [inline containers.go:420]: I am not entirely familiar with the code but I am surprised we're deleting something. I'd imagine something similar to:
```golang
container.Names = dedupeNames(append(containers.Names, names...))
```

**Review by vrothberg — COMMENTED** [review]
Not at all a thorough review

File: containers.go
Comment (vrothberg) [reply containers.go:420]: Scratch that: having read more code, we need to delete the previous references to other containers.

**Review by mtrmac — COMMENTED** [review]
I’d prefer the `struct` approach from https://github.com/containers/storage/pull/1153#pullrequestreview-891044945 because it’s much easier to embed into a larger operation, as discussed above.

But I’ll completely defer to c/storage maintainers.

File: containers.go
Comment (flouthoc) [reply containers.go]: Oh my bad , there was a flag. I forgot to add it.

File: containers.go
Comment (flouthoc) [reply containers.go:406]: Yes this can be done. Just like we do in store.

Comment (mtrmac) [reply containers.go:406]: BTW there’s also the option of the _external_ API having separate `Add`/`Remove`/`Set`, but the three internal stores having an `Edit` with a struct that describes all three.

Up to c/storage maintainers.

File: images.go
Comment (mtrmac) [reply images.go:588]: AFAICS it _only_ adds the newly-set names; not the names that are being removed by this operation, nor the names that have been in the history in the distant past (and not even in the pre-removal `.Names`).

Comment (flouthoc) [reply images.go:588]: We pull names from current image and then re-write the ones which are only needed. I also added a test to verify this behavior.

Comment (flouthoc) [reply images.go:588]: @mtrmac got it let me rectify the test itself and remove the cleanup of history. I'll verify and re-open PR from draft.

Comment (flouthoc) [reply images.go]: fixed.

File: containers.go
Comment (flouthoc) [reply containers.go:392]: Did it with `enum` just like i did it for store.

File: containers.go
Comment (flouthoc) [reply containers.go]: @vrothberg Is there anything for me to do here ? or could you mark resolved.

File: store.go
Comment (flouthoc) [reply store.go]: implemented in below commit.

File: store.go
Comment (flouthoc) [reply store.go:40]: It is being used by other stores as well.

File: store.go
Comment (mtrmac) [inline store.go:2073]: I’m not sure whether to go there — others, feel free to say this is not the time.

If we are adding a new top-level API, do we _have_ to follow the generic `SetNames` model, where the code needs to lock all three stores and guess which one the user meant? I’d expect the user to _always_ know whether the goal is to set a name of a container/layer/image. So we could have `Add{Container,Image,Layer}Names`, that would actually be more efficient than the API provided now, and the tedious `switch` statements in `store.updateNames` would not be necessary.

**Review by mtrmac — COMMENTED** [review]
The `containers.go` comments apply to the other two equally.

File: containers.go
Comment (flouthoc) [reply containers.go:392]: fixed

File: store.go
Comment (flouthoc) [reply store.go:2073]: fixed

File: containers.go
Comment (flouthoc) [reply containers.go]: fixed.

File: containers.go
Comment (flouthoc) [reply containers.go]: Nice idea, I have implemented similar in new commit.

File: containers.go
Comment (mtrmac) [inline containers.go:87]: Please follow https://github.com/golang/go/wiki/Deprecated  in these notices.

File: utils.go
Comment (mtrmac) [inline utils.go:54]: I don’t think the case-folding behavior of `stringutils.InSlice` is expected here.

File: images_test.go
Comment (mtrmac) [inline images_test.go:99]: (Even the `.Len` checks can be dropped, the deep equality comparison also implies the same length, AFAIK.)

**Review by mtrmac — COMMENTED** [review]
Implementation mostly LGTM. I’ll defer to c/storage experts on the (public and internal) API.

**Comment by giuseppe** [comment]
do you also have a patch for buildah that uses these new APIs?

It would be helpful to validate the change and see if it fixes the race condition before we merge

**Comment by flouthoc** [comment]
> do you also have a patch for buildah that uses these new APIs?
> 
> It would be helpful to validate the change and see if it fixes the race condition before we merge

@giuseppe I think new API has to go in `c/common/libimage. Tag` no patch in buildah is needed AFAICS.

File: utils.go
Comment (flouthoc) [reply utils.go]: using custom loop here, stringutils seems to be case insensitive, maybe later we can add a new function in stringutils itself.

File: utils.go
Comment (mtrmac) [inline utils.go:55]: `nameExistsInOldNames` is guaranteed true by the loop containing this line.

Isn’t this variable actually `nameShouldBeRemoved`, or `nameExistsInNames`?[1]


([1] an illustration for the rationale of naming the input value `opParameters`, to make it distinct from all the “names” variables around.)

Comment (flouthoc) [reply utils.go]: I also think `nameShouldBeRemoved` is better, and switched `names` -> `opParameters` for better readability.

**Comment by flouthoc** [comment]
@containers/podman-maintainers @containers/storage-maintainers Following PR and use-case is verified here: https://github.com/containers/podman/pull/13339

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by vrothberg** [comment]
`podman/buildah untag $IMAGE` without a tag will remove all tags/names from the specified image.  To prevent race conditions there, we'd need means to remove all names from an image.

Could we do this in case of `RemoveNames(id, nil)` (i.e., when the argument is `nil`)?

**Comment by flouthoc** [comment]
@vrothberg They can still use `SetNames` for that, cause deleting all names will anyways produce race for other builds.

**Comment by vrothberg** [comment]
> @vrothberg They can still use `SetNames` for that, cause deleting all names will anyways produce race for other builds.

`SetNames` is deprecated, so I don't think any caller should continue using it.

**Comment by flouthoc** [comment]
@vrothberg How did we used to solve this with `SetNames` did we used to pass empty array to `SetNames` ?

**Comment by vrothberg** [comment]
> @vrothberg How did we used to solve this with `SetNames` did we used to pass empty array to `SetNames` ?

Currently, we're first reading all names, and then untag them one-by-one.  But I think we can continue doing that what we're doing and leave this PR as is.  

LGTM

**Comment by giuseppe** [comment]
@nalind are you fine to merge the last version?

**Comment by mtrmac** [comment]
> Could we do this in case of `RemoveNames(id, nil)` (i.e., when the argument is `nil`)?

(I understand we are not planning to do this, so just a side note:) That would be a rather irregular behavior, and potentially a trap to a caller that builds the list of names from components.

We are not going to run out of strings, so adding a special `RemoveAllImageNames` (should that even include digested references?) is always an option.

**Review by mtrmac — COMMENTED** [review]
Code LGTM, I’ll defer to others on the API, e.g. (https://github.com/containers/storage/pull/1153#discussion_r813256132 https://github.com/containers/storage/pull/1153#discussion_r813273185 ). From my POV both of these ideas are not a reason to block this PR.

File: containers.go
Comment (nalind) [reply containers.go:96]: I'm not attached to the new methods being public.  The early guess was that we'd find ourselves needing to call into LayerStore/ImageStore/ContainerStore methods directly far more often than we do now, but I don't think that's turned out to be the case in practice.

File: utils.go
Comment (nalind) [inline utils.go:44]: If all of the name manipulation logic ends up running through here, it's a good place to do the deduplication.

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**
